flytekit/core/type_engine.py

```diff
@@ -1866,6 +1877,9 @@ async def async_to_python_value(
         # This is serial, not actually async, but should be okay since it's more reasonable for Unions.
         for v in get_args(expected_python_type):
             try:
+                if found_res and v is NoneType:
+                    continue
+
```

Comment:
When the input is a msgpack idl, and the expected python type is an Optional.
`trans.async_to_python_value(ctx, lv, int)` and
`trans.async_to_python_value(ctx, lv, None)` both works.
This will cause ambiguity, so I use `continue` to skip this. 

For example:
```python
@dataclass 
class DC:
       a: Optional[int]

@workflow
def wf(dc: DC):
      t_optional_int(a=a)
```